packages/core/nano/src/types.ts

```diff
@@ -73,13 +73,17 @@ export interface IVectorProducer<T> {
 
 export interface IMatrixConsumer<T> {
     /** Notification that rows have been inserted, removed, and/or replaced in the given matrix. */
-    rowsChanged(row: number, numRemoved: number, rowsInserted: T[], producer: IMatrixProducer<T>): void;
+    rowsChanged(row: number, numRemoved: number, numInserted: number, producer: IMatrixProducer<T>): void;
 
     /** Notification that cols have been inserted, removed, and/or replaced in the given matrix. */
-    colsChanged(col: number, numRemoved: number, colsInserted: T[], producer: IMatrixProducer<T>): void;
+    colsChanged(col: number, numRemoved: number, numInserted: number, producer: IMatrixProducer<T>): void;
 
-    /** Notification that a range of cells have been replaced in the given matrix. */
-    cellsReplaced(row: number, col: number, numRows: number, numCols: number, values: T[], producer: IMatrixProducer<T>): void;
+    /**
+     * Notification that a range of cells have been replaced in the given matrix.  If the source
+     * matrix has the new cell values already in an array, it may optionally pass these to consumers
+     * as an optimization.
+     */
+    cellsChanged(row: number, col: number, numRows: number, numCols: number, values: ReadonlyArray<T> | undefined, producer: IMatrixProducer<T>): void;
```

Comment:
Assuming that `values` are row major order?

Reply:
For now.  I think in the future, this might turn into something more flexible.